[Beck] 드로잉 앱 1단계Ⓜ️ 수정 진행 중

SangHwi-Back opened the pull request:
## 이번 주의 작업

이번 PR 은 이전의 구조를 변경하기 위한 작업이 주를 이뤘습니다. 결과는 아래와 같습니다.

1. 구조를 전부 바꾸면서 기존의 기능은 전부 작동하지 않습니다.
2. 유즈케이스를 의존성 주입하는 것을 시작으로 화면의 로직이 동작합니다.
3. SOLID 를 기준으로 앱의 기능을 구현 중입니다.

기능을 전부 구현하지는 못했지만 유즈케이스, 그리고 그를 돕는 여러 객체의 관계를 구상해보며 효율적인 아키텍처를 구현하고 있습니다.

## 변경된 앱의 구조

<img width="704" alt="스크린샷 2023-11-15 오후 11 35 31" src="https://github.com/butterfly-architecture-2023/drawing-app/assets/65931336/c1ea1027-47b0-4392-bb26-34933c601efb">

유즈케이스에 정의된 메소드들은 모델에 어떤 작업을 수행한다는 의미보다는 뷰의 어떤 이벤트가 감지하기 위해 정의되어 있습니다.

각각의 책임을 정확히 하고자 뷰 컨트롤러의 역할을 추상화 하였습니다.
* SRP 를 지킨 덕에 수정해야 할 부분이 명확해 졌습니다.
* 프로토콜을 이용하면서 자동으로 OCP 가 용이해졌습니다.
* LSP 를 이용해 참고해 어떤 유즈케이스던 작성할 수 있게 되었습니다.
* MainViewGestureDelegate 는 ISP 를 지키지 않고 있던 뷰 컨트롤러의 제스처 처리 부분을 정리하기 위해 생성하였습니다.
* DIContainer 는 DIP 를 참고하여 유즈케이스를 의존성 주입하는 객체로 정의하였습니다.

> 여러모로 구현이 부족한 상황입니다. 제출일이 지나도 차근차근 짬을 내어 추가 업로드를 계속 진행하겠습니다.

SangHwi-Back committed:
Resolve build error in Tests

SangHwi-Back committed:
Edit interface of manager

SangHwi-Back committed:
Add Protocols of elements

- 책임, 구성요소를 나누기 위해 프로토콜을 선언하여 기능 세분화

SangHwi-Back committed:
Add Types for architectures

- 유즈케이스, 팩토리, 매니저를 추가하여 앱 기능 세분화 진행

SangHwi-Back committed:
Implement specifics in manager and factory

SangHwi-Back committed:
Define useCase's interface

- 유즈케이스의 인터페이스를 정의. 단순히 유저 액션만을 관찰하고 Manager, Factory 에게 작업 요청

SangHwi-Back committed:
Moving view logic to own useCase

SangHwi-Back committed:
Remove view responsibility

- 제스처 책임을 지웁니다.
- 버튼의 처리를 관찰하는 것으로 변경합니다.

godrm commented on DrawingApp/DrawingApp/Feature/MainViewFactory.swift:
```diff
@@ -12,6 +12,9 @@ class MainViewFactory {
+  
   let manager: DrawingResourceManager
   
   init(width: CGFloat, height: CGFloat) {
```
뷰 팩토리가 생성 역할만 하는 건 아닌가요? 속성이 과하게 많은 것 같은 느낌입니다. 그만큼 책임이 커서 하는 일이 많은걸까요?

godrm commented on DrawingApp/DrawingApp/Feature/MainViewFactory.swift:
```diff
@@ -74,35 +77,54 @@ class MainViewFactory {
+    let set = color.colorSet
+    self.init(red: set.red, green: set.green, blue: set.blue, alpha: set.alpha)
+  }
+}
```
여기에서만 필요한 extension 인 것 처럼 보이네요. 좋습니다

godrm commented on DrawingApp/DrawingApp/Feature/MainViewGestureDelegate.swift:
```diff
@@ -0,0 +1,80 @@
+
+import UIKit
+
+class MainViewGestureDelegate: NSObject, UIGestureRecognizerDelegate {
```
이름에 프로토콜 역할과 동일하게 `-Delegate`를 붙이면 프로토콜인지 구체타입인지 구분이 안될 것 같습니다
추상 타입과 구체 타입 이름 규칙도 정해보면 좋겠습니다

godrm commented on DrawingApp/DrawingApp/Feature/MainViewGestureDelegate.swift:
```diff
@@ -0,0 +1,80 @@
+        useCase?.endVectorGesture(data: data)
+      }
+      imageView = nil
+      lastPoint = nil
```
이벤트를 받아주는 델리게이트 객체까지는 UIKit에 의존한다고 하면, 여기서 구조적인 판단을 내려야 합니다
이후에 있는 UseCase에게 무엇을 전달해 줄 것인가. 뷰에서 보낸 이벤트를 받아서 `Action` 이라는 추상화하는 대상이 필요하고, 세부적으로는 데이터 값이 필요합니다. 
canvas에 그림을 그리더라도 액션을 받아서 바로 그리는 게 아니라 더 하위로 전달하는 것과 분리해보는 것을 추천드립니다

godrm commented on DrawingApp/DrawingApp/Feature/Manager/SquareManager.swift:
```diff
@@ -0,0 +1,42 @@
+  
+  enum Action: String {
+    case highlight, move
+  }
```
이런 액션이 어디서부터 어디까지 전달되는 것인가
액션에 따르는 데이터는 무엇인가도 좀 더 고민해보세요. 액션만 있고 관련된 데이터가 분리되어 있는 느낌이네요

godrm commented on DrawingApp/DrawingApp/Feature/Usecase/MainViewUseCase.swift:
```diff
@@ -0,0 +1,64 @@
+    if let square = square {
+      let _ = squareManager.makeChange(id, action: square.isHighlighted ? .move : .highlight)
+    }
+  }
```
유즈케이스가 Square와 Vector 를 꼭 모두 처리해야 하는 건 아닙니다. 
이것도 선택사항이지만 관련성이 높은 것들만 한꺼번에 처리는 게 좋습니다

godrm commented on DrawingApp/DrawingApp/Feature/iOSMainViewController.swift:
```diff
@@ -6,24 +6,49 @@
-  private var lastPoint: CGPoint?
+  private lazy var gestureDelegate = MainViewGestureDelegate(useCase: useCase, canvas: view)
+  
+  init(useCase: MainViewUseCase) {
+    self.useCase = useCase
+    super.init(nibName: nil, bundle: nil)
+    
+    useCase.objectWillChange.sink { _ in
+      for s in useCase.squares {
+        
+      }
+      for v in useCase.vectors {
+        
+      }
+    }
+    .store(in: &subscriptions)
```
여길 보면 UseCase가 ViewController 다음 계층인 것 같은데 UIKit에 독립적으로 동작하는 것처럼 보이기도 하네요
다만 UseCase가  square와 vector 생성에만 촛점이 있는 것 같습니다. 
그 외에 터치하거나 선택한 이후에 다른 이벤트도 유즈케이스로 만들어보길 추천드립니다. 그래야 구조가 달라질 것 같습니다

godrm commented on DrawingApp/DrawingAppBusinessDomain/DrawingAppBusinessDomain/CanvasSize.swift:
```diff
@@ -8,8 +8,8 @@
+  @RangeFromZero public var height: Double
   
   public init(width: Double, height: Double) {
     self.width = width
```
이런 타입에 property wrapper로 제한을 두더라도 생성 시점에 오류를 인지하기는 어렵습니다. 이 부분도 고려해보세요

godrm reviewed (approved):
안녕하세요 JK입니다
개선하신 방향이 UIKit에 독립적이기는 하지만, UseCase가 확장가능한가 생각하면 그렇지 않은 부분도 있어 보입니다.

DIContainer를 두신 시도는 좋습니다만, 잘못하면 거기가 모든 하위 타입이 모여있는 곳이 되면서 오히려 의존성이 높아질 수 있습니다. 그 점도 고민해보세요 

다음은 제가 미션 설계하면서 고민했던 내용입니다. 그룹에서 같이 의논해보셨으면 좋겠습니다

- 네트워크를 담당하는 모듈은 상위 모듈일까요? 하위 모듈일까요?

- 사각형 View가 사각형 Model을 의존하는지. 아니면 서로 독립적인지

- 1주차처럼 View를 제외하고 핵심 도메인을 구현하고 검증하려고 시도해보셨나요

- SRP 단일 책임을 의식해서 (또는 의도해서) 구현한 객체가 있나요

- DIP 의존성 역전을 위해서 어느 객체에 어떤 시도를 했나요

- 테스트 코드를 작성하면서 의도했던 방향, 검증하려했던 내용은 주로 어떤 역할인가요

SangHwi-Back committed:
Replace new structure

- 뷰 모델을 앞세워 뷰를 모델링 합니다.
- 뷰 모델은 아키텍처에서 비즈니스 로직을 담당하는 유즈케이스, 팩토리와 소통합니다.
- 유즈케이스, 팩토리는 엔티티와 소통합니다.

godrm commented on DrawingApp/DrawingApp/Feature/MainViewModel.swift:
```diff
@@ -0,0 +1,67 @@
+
+import Foundation
+import DrawingAppBusinessDomain
+import UIKit
+
+class MainViewModel: NSObject {
+  private let squareSelectUseCase: SelectSquareElementUseCase
+  private let vectorSelectUseCase: SelectVectorElementUseCase
```
뷰 모델부터는 UIKit에 독립적으로 작성해보시는 것도 시도해보세요
